### PR TITLE
Backend switching

### LINKED ISSUE #3679: Runtime detection for default backend
While I'd like to make every backend available for a distribution package, the Qt5Agg backend will be set as default by matplotlib's installation script, and when a user doesn't have qt5 installed, matplotlib will throw an error if the client application didn't specify a backend.

It would be nice if there could be a fallback logic on runtime so the best-available GUI backend will be chosen as the default backend.

### PR DESCRIPTION
**Pull request by anntzer**
This is the followup to #9551 (backend loading refactor), and includes it (so it would make sense for #9551 to be merged first).

This PR removes the ambiguous behavior of `matplotlib.use`, which was, up to now, documented as follows:
> This function must be called *before* importing pyplot for the first time; or, if you are not using pyplot, it must be called before importing matplotlib.backends.  If warn is True, a warning is issued if you try and call this after pylab or pyplot have been loaded.  "In certain black magic use cases, e.g. pyplot.switch_backend, we are doing the reloading necessary to make the backend switch work (in some cases, e.g., pure image backends) so one can set warn=False to suppress the warnings.)

Instead, matplotlib.use now works just fine as long as an event loop has not started yet (i.e. an interactive window popped up), or when switching to a noninteractive backend, or to an interactive backend compatible with the current event loop (e.g. gtk3agg -> gtk3cairo); in all other cases, an exception is raised.  This means, for example, that one can write `from matplotlib import pyplot as plt; plt.rcParams["backend"] = "foo"; ...` instead of having to do a separate import of matplotlib.

Moreover, it is now possible to set rcParams["backend"] to a list of candidate backends; they will simply be tested one at a time until one succeeds.  If pyplot is already loaded, the value in rcParams["backend"] is immediately replaced by the successful candidate.  If pyplot is not loaded yet, rcParams["backend"] will keep hold of the list of candidates until pyplot is loaded.  This is a documented change of behavior (users should import pyplot to force backend resolution).  (Alternatively the list could be shoved into a new, separate rcParam, but little would be gained because rcParams["backend"] would then hold an incorrect value until backend resolution).  Another change of behavior is that the matplotlib.backends.backend global variable has been removed, again in favor of rcParams["backend"] as single truth point: it was too difficult to properly keep matplotlib.backends.backend in sync, especially at import time when that module may not have been imported yet.

(Exception: I don't have a Mac available to figure out the relevant fixes for the OSX backend, but hopefully another contributor can help here.  See _get_current_event_loop.)

Edit: Now also working on OSX, see comment below.

A major advantage of such a scheme is that we can now just ship a default matplotlibrc which lists all buitin backends in preference order (`backend: osx, qt5agg, qt4agg, gtk3agg, ...`).  Currently, this list is evaluated at build time (by testing which GUI toolkit is importable), which causes two major problems: 1) the import of the GUI toolkits need to happen in separate processes, which means that setup.py relies on multiprocessing(!); for example, this appears to confuse conda-build (see comment in ci/conda_recipe/meta.yaml); 2) the built list is useless anyways unless 1) the end user is doing the build, rather than a central packager (the end user may have a different list of installed GUI toolkits...) and 2) the end user never adds or removes a new GUI toolkit.  Here, I can instead completely strip out the relevant part of setupext.py, greatly simplifying it.

The backend_fallback rcParam likewise becomes obsolete, although this PR does not remove it (this rcParam can easily go through a normal deprecation route).

I also removed the interactive backends API docs because the shimming (to allow them to be imported in a headless buildbot) was insufficient to support this PR's implementation.  Proposed fix at https://github.com/matplotlib/matplotlib/pull/9708#issuecomment-342592846.  (But note that they were incomplete anyways...)

Closes #3679, #6739, #8613, https://github.com/matplotlib/matplotlib/issues/9017#issuecomment-349868494; sort-of #3466 (which is perhaps more a doc issue...).

Some comments to try clarifying the implementation:
- matplotlib.use only supports a single argument (not a list of candidate backends) because we want to keep `use(...); rcdefaults()` working: this implies that `use(...)` must overwrite the default backend (in rcParamsDefault, and likewise rcParamsOrig), to a single fixed value.
- setting multiple entries in rcParams["setup"] defers to pyplot.switch_backends, which itself defers to backends.pylab_setup.  Once the correct candidate is found, backends.pylab_setup sets the single entry in rcParams["setup"].

### CODE REVIEW

**Comment by jklymak** [comment]
This looks *really* cool.  But, I'm on MacOS so its not working for me yet.  If I do 

```python
import matplotlib
print(matplotlib.get_backend())
matplotlib.use('Qt5Agg')
import matplotlib.pyplot as plt
plt.scatter([1, 2], [0,2])
print(matplotlib.get_backend())
plt.show()
```

I get a Qt5Agg GUI and
```
['MacOSX', 'Qt5Agg', 'Qt4Agg', 'GTK3Agg', 'GTK3Cairo', 'GTKAgg', 'TkAgg', 'WXAgg', 'agg', 'cairo']
Qt5Agg
```

If I do 

```python
import matplotlib
print(matplotlib.get_backend())
import matplotlib.pyplot as plt
matplotlib.use('Qt5Agg')
plt.scatter([1, 2], [0,2])
print(matplotlib.get_backend())
plt.show()
```
I get:

['MacOSX', 'Qt5Agg', 'Qt4Agg', 'GTK3Agg', 'GTK3Cairo', 'GTKAgg', 'TkAgg', 'WXAgg', 'agg', 'cairo']
Qt5Agg

**but**, I get the macOS GUI (Toolbar on the bottom instead of top, and I can close with command-w).  So either I am not understanding how to test this or MacOS needs work.

Happy to help make this work on MacOS...

**Comment by anntzer** [comment]
That's because I forgot to actually update the backend when doing `rcParams["backend"] = "foo"` (as opposed to `use()`), should hopefully be fixed now (if I didn't break something else in the meantime, but at least the example you posted works for me replacing OSX/Qt5 by Gtk3/Qt5).

Right now what needs to be done for OSX is a way to detect whether the OSX event loop is already running (see _get_current_event_loop for examples), so that having an OSX event loop running prevents starting another event loop.

**Comment by jklymak** [comment]
Confirmed, the snippet above now works as advertized!

**Comment by jklymak** [comment]
BTW, looking at event-loops and such would be a longer term project for me if I were to help.  @dopplershift would probably be much better suited.

**Comment by anntzer** [comment]
Checking whether the NSApp event loop is running turned out to be quite easy (again, thanks @jklymak).

With this PR, on a default OSX install, running `python -c 'from pylab import *; plot(); show()'` (instead of python.app) now automatically falls back on qt5agg (and should do so all the way down to tkagg if pyqt5 is not available, but that I haven't tested) instead of displaying the "Python is not installed as a framework" error.  Of course, if the backend is forced to macosx (via MPLBACKEND or matplotlib.use or matplotlibrc) then the error is still displayed.

File: lib/matplotlib/backends/__init__.py [hunk 1]
```diff
@@ -44,53 +87,60 @@ def pylab_setup(name=None):
...
-
-    show = getattr(backend_mod, 'show', do_nothing_show)
-
-    draw_if_interactive = getattr(backend_mod, 'draw_if_interactive',
-                                  do_nothing)
-
-    _log.info('backend %s version %s' % (name, backend_version))
+        name = matplotlib.rcParams["backend"]
+
+    if isinstance(name, list):
```
Comment (tacaswell) [inline lib/matplotlib/backends/__init__.py]: testing if not `str` and and iterable is probably better (so we support anything that ducks as a list here.

**Comment by tacaswell** [comment]
:+1: in principle, have not given it a careful read yet.

File: lib/matplotlib/backends/__init__.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/backends/__init__.py]: changed to "not str".

**Comment by jklymak** [comment]
@anntzer I can't get the code snippet above to fall back to PyQT5; I still get the framework error.  Pretty sure I don't have a matplotlibrc installed or have $MPLBACKEND set.

**Comment by anntzer** [comment]
I just tried on your machine and it worked for me...  Can you check what's happening at INFO log level, especially during the call to matplotlib.backends.pylab_setup?

**Comment by jklymak** [comment]
@anntzer is this still active?  Sorry if I was supposed to get back to you on some of it - looks like it'll be an ugly rebase, but they are probably all your changes anyways :wink:

**Comment by anntzer** [comment]
I'll probably wait until the rest of the mpl3 transition settles down a bit before revisiting this.

**Comment by tacaswell** [comment]
Some version of this should go in for 3.0 to make life easier for conda-forge.

**Comment by tacaswell** [comment]
Closing in favor of #11519, #11520, and #11600

**Pull request closed without merging**